File: src/hanlint/report/operationMatch.py
"""현재 문장에 승인 표면 치환이 안전하게 하나만 적용되는지 계산한다."""

from __future__ import annotations

from collections.abc import Iterable

from ..data import Patch, SurfaceOperation, operationFor
from ..fingerprint import DocumentPrint, sourceSentenceTexts
from ..rules import Finding
from .patchMatch import selectedPatch


def operationGuidance(
    doc: DocumentPrint | None,
    findings: Iterable[Finding],
    preset: str | None,
    operations: Iterable[SurfaceOperation],
    patches: Iterable[Patch] = (),
    protectedTerms: Iterable[str] = (),
) -> list[dict]:
    """확정 fix나 원문 완전 일치 패치가 맡은 문장은 건너뛰고 유일한 표면 치환만 싣는다."""
    operations = tuple(operations)
    findings = tuple(findings)
    patches = tuple(patches)
    if doc is None or not preset or not operations:
        return []
    reserved = {finding.at for finding in findings if finding.at >= 0 and finding.fix is not None}
    reserved.update(
        finding.at for finding in findings if finding.at >= 0 and selectedPatch(doc, finding, preset, patches) is not None
    )
    sourceTexts = sourceSentenceTexts(doc)
    guidance = []
    for sentence in doc.sentences:
        if sentence.index in reserved:
            continue
        sourceText = sourceTexts.get(sentence.index)
        if sourceText is None:
            continue
        applied = operationFor(sourceText, preset, operations, protectedTerms)
        if applied is not None:
            guidance.append({"line": sentence.line, "operation": applied.asDict(preset)})
    return guidance

File: npm/src/report/operationMatch.js
// @ts-check
/** 현재 문장에 승인 표면 치환이 안전하게 하나만 적용되는지 계산한다. */
import { operationData, operationFor } from "../data/operations.js";
import { sourceSentenceTexts } from "../fingerprint/sourceSentence.js";
import { selectedPatch } from "./patchMatch.js";

/**
 * @param {import("../fingerprint/build.js").DocumentPrint | null | undefined} document
 * @param {import("../rules/finding.js").Finding[]} findings
 * @param {string | null | undefined} preset
 * @param {import("../data/operations.js").SurfaceOperation[]} operations
 * @param {import("../data/patches.js").Patch[]} [patches]
 * @param {string[]} [protectedTerms]
 */
export function operationGuidance(document, findings, preset, operations, patches = [], protectedTerms = []) {
  if (!document || !preset || !operations.length) return [];
  const reserved = new Set(findings.filter((finding) => finding.at >= 0 && finding.fix !== null).map((finding) => finding.at));
  for (const finding of findings) {
    if (finding.at >= 0 && selectedPatch(document, finding, preset, patches)) reserved.add(finding.at);
  }
  const sourceTexts = sourceSentenceTexts(document);
  const guidance = [];
  for (const sentence of document.sentences) {
    if (reserved.has(sentence.index)) continue;
    const sourceText = sourceTexts.get(sentence.index);
    if (sourceText === undefined) continue;
    const applied = operationFor(sourceText, preset, operations, protectedTerms);
    if (applied) guidance.push({ line: sentence.line, operation: operationData(applied, preset) });
  }
  return guidance;
}
